### ISSUE TITLE
[CT-2997] Ability to test incremental models

### ISSUE THREAD

**Issue by gshank** [dbt-labs/dbt-core#8422]
### Description

We should allow folks to add unit tests to incremental models

### Acceptance criteria

* Ability to test incremental models in both full refresh and incremental mode
* Ability to specify the current status of an incremental model to test the incremental behavior

More on the above two points in https://github.com/dbt-labs/dbt-core/discussions/8275#discussioncomment-6825650: 
> To support this, dbt's unit testing framework would need to provide the ability to:
* override the is_incremental macro to return True (p0), and provide the ability to specify an input fixture for {{ this }} (p1) in order to mock out the previous state of an incremental model. The expect output would be the set of rows that dbt's incremental materialization would insert/merge into the existing relation in production, as opposed to the resulting updated representation of the model.

To be moved into a separate issue: 
* Ability to test the incremental behavior of models using incremental strategy that requires partitioning (e.g. insert overwrite in BigQuery): https://github.com/dbt-labs/dbt-core/issues/8664 (blocked on spike #8499)

A "current_actual" could define the current state of the model we're incrementing on, testing whether the incrementing works as expected.

**Comment by martynydbt** [comment]
@gshank please pull Ability to test the incremental behavior of models using incremental strategy that requires partitioning (e.g. insert overwrite in BigQuery) out into a separate ticket

**Comment by jtcohen6** [comment]
2x2 = 4 acceptance tests for us to write as part of implementation:
- {{ is_incremental() }} : True + False
- {{ this }} fixture: provided or not

**Comment by MichelleArk** [comment]
Here's how i'm imagining the spec for `this`:
```
unit-tests: 
 - model: my_model
   tests: 
    - name: test_name
      given: 
      - input: this
         rows: ... 
```
